### DISCUSSION TITLE
Is mutex in the Pipeline struct necessary?

### DISCUSSION THREAD

**Discussion by ash2k** [redis/go-redis#2351] [General]
https://github.com/go-redis/redis/blob/86f4ea13a97aef79e108c428ec121fed0d60f1d1/pipeline.go#L43

Why would someone want to use a pipeline concurrently? Seems weird. Most people pay the price of the mutex without a need for it (99%?). People who actually need to use it concurrently (1%?) can just wrap it in a mutex by hand.

I suggest to remove the mutex while v9 hasn't been released yet.

**Comment by vmihailenco** [comment]
@monkey92t any thoughts? I don't have any preferences here...

**Comment by monkey92t** [comment]
Looks like yes, I believe for 100% (not 99% 😁) of people, pipeline doesn't need mutex, we can remove it, but need to warn users that pipeline is not concurrency safe.

**Comment by monkey92t** [comment]
Also....Should we allow reuse of `Pipeline`?

**Reply by vmihailenco** [reply]
@monkey92t I think yes, Pipeline should be reusable. I already rely on that [here](https://github.com/go-redis/redis/blob/master/example/del-keys-without-ttl/main.go#L132-L154)

**Comment by vmihailenco** [comment]
https://github.com/go-redis/redis/discussions/2351
